File: checks/python/src/section508/catalog.py
from functools import cache
from pathlib import Path

import yaml

CATALOG_PATH = Path(__file__).resolve().parents[3] / "catalog.yml"


@cache
def load_catalog() -> dict:
    with CATALOG_PATH.open(encoding="utf-8") as handle:
        return yaml.safe_load(handle) or {}


def rules() -> list[dict]:
    return list(load_catalog().get("rules") or [])


def rules_by_id() -> dict[str, dict]:
    return {rule["id"]: rule for rule in rules() if "id" in rule}


def rule_ids(*, rules=None, profile=None) -> list[str]:
    catalog = rules_by_id()
    if rules:
        ids = list(rules)
    else:
        ids = list(catalog.keys())
    if profile:
        ids = [
            rule_id
            for rule_id in ids
            if profile in ((catalog.get(rule_id) or {}).get("profiles") or [])
        ]
    return ids

File: checks/ruby/lib/section508/catalog.rb
# frozen_string_literal: true

require "yaml"

module Section508
  module Catalog
    PATH = File.expand_path("../../../catalog.yml", __dir__)

    module_function

    def data
      @data ||= YAML.safe_load_file(PATH) || {}
    end

    def rules
      Array(data["rules"])
    end

    def by_id
      @by_id ||= rules.each_with_object({}) { |rule, map| map[rule["id"]] = rule }
    end

    def rule_ids(rules: nil, profile: nil)
      ids = if rules.nil? || (rules.respond_to?(:empty?) && rules.empty?)
              self.rules.map { |rule| rule["id"] }
            else
              Array(rules).map(&:to_s)
            end
      if profile
        ids = ids.select { |id| Array(by_id.dig(id, "profiles")).include?(profile) }
      end
      ids
    end
  end
end
